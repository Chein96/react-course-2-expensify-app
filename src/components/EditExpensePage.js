import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import ExpenseForm from './ExpenseForm';
import { startRemoveExpense, startEditExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    state = {
        visibility: false
    }

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.setState(()=>({visibility:false}));
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    changeVisibility = () => this.setState((prevState)=>({visibility: !prevState.visibility}));

    render(){
        const { visibility } = this.state;

        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.changeVisibility}>Remove Expense</button>
                </div>
                <Modal 
                    isOpen={visibility}
                    onRequestClose={this.changeVisibility}
                    contentLabel="Expense removal confirmation"
                    closeTimeoutMS={200}
                    className="modal"
                >
                    <h3 className="modal__title">
                        <span>"{this.props.expense.description}"</span>
                        will be removed.
                    </h3>
                    <div className="modal__actions">
                        <button className="button" onClick={this.onRemove}>Continue</button>
                        <button className="button button--secondary" onClick={this.changeVisibility}>Cancel</button>
                    </div>
                </Modal>
            </div>
        )
    }
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: ({id}) => dispatch(startRemoveExpense({id}))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);