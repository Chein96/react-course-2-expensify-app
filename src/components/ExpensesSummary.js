import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal, hiddenExpensesCount }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const hiddenExpenseWord = hiddenExpensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,00.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span>
                </h1>
                <h3 className="page-header__subtitle">
                    {
                        hiddenExpensesCount !== 0 ? (
                            `${hiddenExpensesCount} hidden ${hiddenExpenseWord} with current filters.`
                        ) : (
                            `No hidden expenses.`
                        )
                    }
                </h3>
                
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps = ({ expenses, filters }) => {
    const visibleExpenses = getVisibleExpenses(expenses, filters);
    const hiddenExpensesCount = expenses.length - visibleExpenses.length;
    return { 
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses),
        hiddenExpensesCount
    }
};

export default connect(mapStateToProps)(ExpensesSummary);