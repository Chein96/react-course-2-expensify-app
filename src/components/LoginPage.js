import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div>
        Welcome to the Expensify App
        <button onClick={startLogin}>Login</button>
    </div>
);

const mapDisaptchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDisaptchToProps)(LoginPage);