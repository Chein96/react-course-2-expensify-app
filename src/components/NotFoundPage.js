import React from 'react';
import { Link } from 'react-router-dom';

export const NotFoundPage = () => (
    <div>
        Error 404.
        <Link to="/">Go home</Link>
    </div>
);

export default NotFoundPage;