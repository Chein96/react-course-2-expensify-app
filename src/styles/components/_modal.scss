.ReactModalPortal > div {
    opacity: 0;
}

.ReactModalPortal .ReactModal__Overlay {
    align-items: center;
    background-color: rgba(0,0,0,0.25) !important;
    display: flex;
    justify-content: center;
    transition: opacity 200ms ease-in-out;
}

.ReactModalPortal .ReactModal__Overlay--after-open {
    opacity: 1;
}

.ReactModalPortal .ReactModal__Overlay--before-close {
    opacity: 0;
}

.modal {
    background: $off-white;
    max-width: 30rem;
    outline: none;
    padding: $l-size;
    text-align: center;
}

.modal__title {
    margin: 0 0 $m-size 0;
    font-weight: 300;

    > span {
        display: block;
        font-weight: 800;
        word-break: break-all;
    }
}

.modal__actions {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    > .button--secondary {
        margin-top: $s-size;
    }

    @media (min-width: $desktop-breakpoint) {
        flex-direction: row;

        > .button--secondary {
            margin: 0 0 0 $s-size;
        }
    }
}