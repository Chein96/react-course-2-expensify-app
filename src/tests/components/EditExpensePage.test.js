import React from 'react';
import { shallow } from 'enzyme';
import Modal from 'react-modal';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let wrapper, history, startEditExpense, startRemoveExpense;
beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage
        history={history}
        startEditExpense={startEditExpense}
        startRemoveExpense={startRemoveExpense}
        expense={expenses[1]}
    />);
});

test('Should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle startEditExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('Should handle changeVisibility on click', () => {
    wrapper.find('button').at(0).simulate('click');
    expect(wrapper.state('visibility')).toBeTruthy();
});

test('Should handle startRemoveExpense on confirmation modal', () => {
    wrapper.find('button').at(1).simulate('click');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('Should hide confirmation modal', () => {
    wrapper.find('button').at(0).simulate('click');
    wrapper.find('button').at(2).simulate('click');
    expect(wrapper.state('visibility')).toBeFalsy();
});

test('Should hide confirmation modal on user request', () => {
    wrapper.find('button').at(0).simulate('click');
    wrapper.find(Modal).prop('onRequestClose')();
    expect(wrapper.state('visibility')).toBeFalsy();
});