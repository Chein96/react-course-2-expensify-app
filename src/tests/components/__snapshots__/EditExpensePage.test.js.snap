// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 109500,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={200}
    contentLabel="Expense removal confirmation"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <h3
      className="modal__title"
    >
      <span>
        "
        Rent
        "
      </span>
      will be removed.
    </h3>
    <div
      className="modal__actions"
    >
      <button
        className="button"
        onClick={[Function]}
      >
        Continue
      </button>
      <button
        className="button button--secondary"
        onClick={[Function]}
      >
        Cancel
      </button>
    </div>
  </Modal>
</div>
`;
